### ISSUE TITLE
`waitFor` proposal

### ISSUE THREAD

**Issue by andre-merzky** [ExaWorks/job-api-spec#73]
The [`waitFor`](https://github.com/ExaWorks/job-api-spec/blob/main/specification.md#job-waitfor) method currently accepts an timeout and a set of target states.  Following a [discussion with Stephen](https://github.com/ExaWorks/jpsi-python/pull/2/files#r546202828), I suggest the following changes

  - rename `waitFor` to `wait` (in sync with other job and process APIs)
  - let the call always return when a final state is reached.

About that last item: assume `job.wait(timeout=None, states=[ACTIVE])`, and the job fails to get queued, so it goes to `FAILED` before ever passing `ACTIVE`.  The call would now hang forever.  In general, waiting on jobs in a final state is rather useless.  We could also consider raising an exception if the job goes into a final state which is not explicitly listed as a target state though.

**Comment by hategan** [comment]
> The [`waitFor`](https://github.com/ExaWorks/job-api-spec/blob/main/specification.md#job-waitfor) method currently accepts an timeout and a set of target states. Following a [discussion with Stephen](https://github.com/ExaWorks/jpsi-python/pull/2/files#r546202828), I suggest the following changes
> 
> * rename `waitFor` to `wait` (in sync with other job and process APIs)

wait() is a final method in Java, so, in Java it can't be wait().

> * let the call always return when a final state is reached.

That makes sense. The purpose of waitFor() is to wait until a condition is reached. There is no point in waiting for something that will never happen.

> 
> About that last item: assume `job.wait(timeout=None, states=[ACTIVE])`, and the job fails to get queued, so it goes to `FAILED` before ever passing `ACTIVE`. The call would now hang forever. In general, waiting on jobs in a final state is rather useless. We could also consider raising an exception if the job goes into a final state which is not explicitly listed as a target state though.

Yes, I think it might make sense to not follow the normal flow when abnormal things happen. So an exception might make sense.

**Comment by andre-merzky** [comment]
> wait() is a final method in Java, so, in Java it can't be wait().

I understand then that the Java binding will have `waitFor`, but that should not affect all other languages, right?

**Comment by hategan** [comment]
> > wait() is a final method in Java, so, in Java it can't be wait().
> 
> I understand then that the Java binding will have `waitFor`, but that should not affect all other languages, right?

It was my impression that we're aiming for a spec that doesn't leave that much room for interpretation.

And I also don't think we should spend time on naming issues unless the offence is particularly grave. That said, whatever...

**Comment by SteVwonder** [comment]
Moved the variation across languages discussion over to #27 to keep this issue targeted.

I too like the proposal to always return when a job reaches a final state, and if the final state reach was not in the waiting list, raise an exception.  Since it sounds like we have consensus there, I propose we table the `wait` vs `waitFor` change until after we have more consensus with #27, but move forward with the other changes.  Sound good?

**Comment by hategan** [comment]
I gave the exception a bit more thought. I don't dislike the idea of `wait(SUBMITTED)` throwing one if the job completes. However, I'd like to address another issue with `wait()` that has been bothering me. It makes a bit less sense with a mostly-linear state transition diagram, but I'd like to make it more robust in case we update that. Consider the scenario in which a user wants to wait for one of two states, then do something depending on which one of those states was reached. The code might look like this:

```python
job.wait(STATE1, STATE2)
state = job.status.state
if state == STATE1:
    doSomething()
else:
    assert(state == STATE2)
    doSomethingElse()
```

The assertion may fail if the job transitions to another state after `wait()` but before `state = ...`. So this way of doing things is inherently raceful. The only way of solving this problem is if the moment `wait()` takes its action and the `state = ...` assignment are protected from changes to `job.status.state` by a critical section. Without exposing internal synchronization details of the implementation, that looks impossible. A better solution is for `job.wait()` to return the state/status that has triggered its completion. E.g.:

```python
status = wait(STATE1, STATE2)
if status.state == STATE1:
    ...
```

However, if we do that, we have a perfectly reasonable mechanism for telling if `wait(STATE)` has returned because the job has reached `STATE` or because it has reached another, final, state. So returning the state looks like a more general solution than throwing exceptions.

**Comment by andre-merzky** [comment]
Wait returning the reached state is nice, I like that.

But not raising on non-specified states is less surprising than possibly returning a state the user has not been waiting for.  I probably could live with that, bit it leaves me a bit uneasy...

I think the two items are not exclusive either: wait could return the reached state, and still raise on final states if those were not explicitly specified.

**Comment by hategan** [comment]
> returning a state the user has not been waiting for. I probably could live with that, bit it leaves me a bit uneasy...

Me too. I think one way out of it would be to clearly indicate this in the signature: 

`wait(additional_states)` - waits for a final state; additional states may be specified with `additional_states`

Or something of that sort.



> I think the two items are not exclusive either: wait could return the reached state, and still raise on final states if those were not explicitly specified.

That, too, makes me a bit uneasy. Perhaps more than the other uneasy. It's because proper handling of calls to `wait()` now require both a conditional check and a `try/except`.

**Comment by andre-merzky** [comment]
>  It's because proper handling of calls to `wait()` now require both a conditional check and a `try/except`.

You only check return value when you are interested in that, like, when you wait for more then one state and need to distinguish them:
```py
state = job.wait()
print('job is final now')
if state == jpsi.FAILED:
    resubmit()
```

You only `try/except` when a final state is unexpected
```py
try:
    job.wait(jpsi.ACTIVE)
    print('job is active now')
except:
   print('could not run job')
```

But you can always get the behaviour you suggest with
```py
state = job.wait(jpsi.FINAL + [ACTIVE])   # will not raise
```

and the default is benign anyway:
```py
job.wait()  # will not raise
```

I guess all options listed in the thread are acceptable one way or the other, they ultimately provide the same semantics, so the decision is more a matter of preference and maybe prevalence of use cases (but we don't know the latter really).

**Comment by hategan** [comment]
> ```python
> state = job.wait(jpsi.FINAL + [ACTIVE])   # will not raise
> ```
> 
> 
> and the default is benign anyway:
> ```python
> job.wait()  # will not raise
> ```

Looks odd. Intuitively, I'd like to describe this about `wait()` as "will only raise if state not specified in param list is reached". But the second doesn't fit that.

The exception is also starting to rub me the wrong way for another reason. It could and probably should be thrown earlier if one can assert that no transition to the desired state is possible. The fact that we pick final states to throw exceptions is just a lazy and incomplete way of  asserting unreacheability.

**Comment by andre-merzky** [comment]
> Looks odd.

Which one?  The latter only defaults to the final states as default for the target states, which the specification doc already says.  I think that default is sensible.  Maybe I am missing the point?

> The exception is also starting to rub me the wrong way for another reason. It could and probably should be thrown earlier if one can assert that no transition to the desired state is possible. The fact that we pick final states to throw exceptions is just a lazy and incomplete way of asserting unreacheability.

Don't be so hard on us, I don't think you/us was/were lazy :-)  But yes, I agree: raising once non of the specified (or implied) states is reachable warrants the exception.

**Comment by hategan** [comment]
> > Looks odd.
> 
> Which one? The latter only defaults to the final states as default for the target states, which the specification doc already says. I think that default is sensible. Maybe I am missing the point?

The inconsistency. The idea of the exception is for it to be raised if the state specified in the exception list cannot be reached. But this isn't so with `wait()`, since any state fits that description, yet no state triggers and exception. I know it's documented, but nobody reads the documentation if they can infer things instead. But that inference will fail here.

I was thinking that one way to make it more consistent is to:
- never throw exceptions (a solution that I would have otherwise liked), but return state that triggers completion of `wait()` instead
- all forms of `wait()` implicitly wait for final states

But I don't like this either, since `wait(states...)`, at least in one interpretation, implicitly waits for all states from which the desired states cannot be reached.

So anyway, let's go with exception, otherwise return state, and without FINAL (final states raise automatically and user can handle, so no need for two ways to do it) and also with recommendation that implementations can as soon as they determine that desired states cannot be reached. Or maybe make that mandatory. Would that work?

**Comment by SteVwonder** [comment]
> So anyway, let's go with exception, otherwise return state, and without FINAL (final states raise automatically and user can handle, so no need for two ways to do it) and also with recommendation that implementations can as soon as they determine that desired states cannot be reached. Or maybe make that mandatory. Would that work?

Just to confirm I understand the proposal:

if a user asks to wait on only `ACTIVE` (i.e., `job.wait(jpsi.ACTIVE)`, but a `FINAL` state is reached before `ACTIVE` (e.g,. `CANCELED`), then that is an exception.  But if a user does `job.wait([jpsi.ACTIVE, jpsi.CANCELED])` and the job reaches `CANCELED` before `ACTIVE`, then no exception and `CANCELED` is returned.  Does that sound right?

**Comment by hategan** [comment]
> if a user asks to wait on only `ACTIVE` (i.e., `job.wait(jpsi.ACTIVE)`, but a `FINAL` state is reached before `ACTIVE` (e.g,. `CANCELED`), then that is an exception. But if a user does `job.wait([jpsi.ACTIVE, jpsi.CANCELED])` and the job reaches `CANCELED` before `ACTIVE`, then no exception and `CANCELED` is returned. Does that sound right?

Yes.

**Comment by hategan** [comment]
Fixed via #98
